### PR TITLE
Quantization and Training of Neural Network for Efficient Inter-Arithmetic-Only Inference(reproduction of the paper with the same name) 

### PR DESCRIPTION
**Pull request by mutewall**
### Checklist
- [x] I've tested that my changes are compatible with the latest version of Tensorflow.
- [x] I've read the [Contribution Guidelines](https://github.com/tensorlayer/tensorlayer/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

 A quantization scheme 
### Description

Jacob, B., Kligys, S., Chen, B., Zhu, M., Tang, M., & Howard, A., et al. (2017). Quantization and training of neural networks for efficient integer-arithmetic-only inference.
paper link: https://arxiv.org/abs/1712.05877

### CODE REVIEW

**Comment by DEKHTIARJonathan** [comment]
Very nice job @mutewall, do you thin you can add the references (research papers) related to your implementation if there is any inside the docstring ?

I would love to read more details about this ;)

Cheers

**Comment by mutewall** [comment]
@DEKHTIARJonathan I have put the link in the "tutorial_quanconv_cifar.py". If it is not obvious enough, where should I add the reference?
And here is the link: https://arxiv.org/abs/1712.05877

**Comment by DEKHTIARJonathan** [comment]
@mutewall Look at this PR: #737, @zsdonghao added directly the reference inside the docstring ;)
Just add the title, authors and link to arxiv paper. It will be perfect :)

**Comment by mutewall** [comment]
@DEKHTIARJonathan I have add the layers in the tests folder. Please check.

**Comment by mutewall** [comment]
@DEKHTIARJonathan I have added the layers in the tests folder. Please check.

**Comment by zsdonghao** [comment]
@mutewall thanks, this PR is almost done. two more things need to be do before you merge it.

1) for consistent file name, tutorial name:  cifar -->  cifar10    
2) to allow more people read your code, add example links to the example page: 
https://github.com/tensorlayer/tensorlayer/blob/master/docs/user/example.rst
https://github.com/tensorlayer/awesome-tensorlayer

**Comment by mutewall** [comment]
@zsdonghao  I have renamed the file and added the example links. Please check.

**Comment by zsdonghao** [comment]
@mutewall cool, this PR is already approved, but you still need to make travis pass before you can click the `merge bottom`.

File: docs/user/example.rst [hunk 1]
```diff
@@ -28,6 +28,7 @@ Vision
  - BinaryNet. Model compression, see `mnist <https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_binarynet_mnist_cnn.py>`__ `cifar10 <https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_binarynet_cifar10_tfrecord.py>`__.
  - Ternary Weight Network. Model compression, see `mnist <https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_ternaryweight_mnist_cnn.py>`__ `cifar10 <https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_ternaryweight_cifar10_tfrecord.py>`__.
  - DoReFa-Net. Model compression, see `mnist <https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_dorefanet_mnist_cnn.py>`__ `cifar10 <https://github.com/tensorlayer/tensorlayer/blob/master/example/tutorial_dorefanet_cifar10_tfrecord.py>`__.
+ - QuanCNN. Model compression, sees `mnist <https://github.com/XJTUI-AIR-FALCON/tensorlayer/blob/master/example/tutorial_quanconv_mnist.py>`__ `cifar10 <https://github.com/XJTUI-AIR-FALCON/tensorlayer/blob/master/example/tutorial_quanconv_cifar10.py>`__.
```
Comment (DEKHTIARJonathan) [inline docs/user/example.rst]: Remove change from here and open a PR here: https://github.com/tensorlayer/awesome-tensorlayer

File: tests/test_layers_convolution.py [hunk 2]
```diff
@@ -258,19 +353,21 @@ def test_layer_n1(self):
...
+        self.assertEqual(self.net2.outputs.get_shape().as_list()[1:],
+                         [299, 299, 64])
 
 
 if __name__ == '__main__':
 
     tf.logging.set_verbosity(tf.logging.DEBUG)
     tl.logging.set_verbosity(tl.logging.DEBUG)
 
-    unittest.main()
```
Comment (DEKHTIARJonathan) [inline tests/test_layers_convolution.py]: Please rollback this change. `unittest.main()` should be inside the `if`

**Comment by luomai** [comment]
@lgarithm This PR also fails because of the import error. Merge with the master does not resolve the issue.

File: docs/user/example.rst (hunk 1 above)
Comment (luomai) [reply docs/user/example.rst]: We can put the links in both documentation and awesome-tl to increase the visibility.

**Comment by luomai** [comment]
This PR LGTM. I am fine with putting the entry point in the documentation.

**Comment by DEKHTIARJonathan** [comment]
@mutewall please update the changelog with all your modifications and we can merge ;)

File: example/tutorial_quanconv_cifar10.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+# train
+n_epoch = 50000
+learning_rate = 0.0001
+print_freq = 1
+batch_size = 128
+
+train_op = tf.train.AdamOptimizer(learning_rate, beta1=0.9, beta2=0.999, epsilon=1e-08,
+                                  use_locking=False).minimize(cost)
+
+tl.layers.initialize_global_variables(sess)
```
Comment (lgarithm) [inline example/tutorial_quanconv_cifar10.py]: This function will be deprecated soon, see
https://github.com/tensorlayer/tensorlayer/blob/master/tensorlayer/layers/utils.py#L198
for alternative solution.

File: example/tutorial_quanconv_mnist.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+cost_test = tl.cost.cross_entropy(y2, y_, name='xentropy2')
+correct_prediction = tf.equal(tf.argmax(y2, 1), y_)
+acc = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
+
+# define the optimizer
+train_params = tl.layers.get_variables_with_name('quan_cnn', True, True)
+train_op = tf.train.AdamOptimizer(learning_rate=0.0001).minimize(cost, var_list=train_params)
+
+# initialize all variables in the session
+tl.layers.initialize_global_variables(sess)
```
Comment (lgarithm) [inline example/tutorial_quanconv_mnist.py]: tl.layers.initialize_global_variables is deprecated in favor of `tf.global_variables_initializer`

File: example/tutorial_quanconv_cifar10.py (hunk 3 above)
Comment (DEKHTIARJonathan) [reply example/tutorial_quanconv_cifar10.py]: +1

File: example/tutorial_quanconv_mnist.py (hunk 4 above)
Comment (DEKHTIARJonathan) [reply example/tutorial_quanconv_mnist.py]: +1

**Comment by DEKHTIARJonathan** [comment]
Thanks for this awesome PR, I merge

**Pull request merged**